File: Coding In Flow/Broadcast Receiver/LocalReceiverSender/app/src/main/java/com/example/localreceiversender/MainActivity.java
package com.example.localreceiversender;

import androidx.appcompat.app.AppCompatActivity;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import android.Manifest;
import android.content.BroadcastReceiver;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {
    private TextView textView;
    private ExampleBroadcastReceiver2 exampleBroadcastReceiver2 = new ExampleBroadcastReceiver2();
    private LocalBroadcastManager localBroadcastManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        textView = findViewById(R.id.textView);

        localBroadcastManager = LocalBroadcastManager.getInstance(this);
    }

    public void sendBroadcast(View view) {
        Intent intent = new Intent("com.example.EXAMPLE_ACTION");
        localBroadcastManager.sendBroadcast(intent);
    }

    @Override
    protected void onStart() {
        super.onStart();
        IntentFilter filter = new IntentFilter("com.example.EXAMPLE_ACTION");
        localBroadcastManager.registerReceiver(exampleBroadcastReceiver2, filter);

    }

    @Override
    protected void onStop() {
        super.onStop();
        unregisterReceiver(exampleBroadcastReceiver2);
    }
}

File: Personal/DataStorage/internal_storage/src/main/java/com/vanphuc/datastorage/MainActivity.kt
package com.vanphuc.datastorage

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import com.vanphuc.datastorage.databinding.ActivityMainBinding
import java.io.*


class MainActivity : AppCompatActivity(), MainActivityListener {
    lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)

        binding.listener = this@MainActivity
    }

    override fun saveData() {
        val fileOutputStream: FileOutputStream?

        try {
            fileOutputStream = openFileOutput(fileName, Context.MODE_PRIVATE)
            fileOutputStream.write(content.toByteArray())
            fileOutputStream.close()
            Toast.makeText(this, "Save data successfully", Toast.LENGTH_SHORT).show()
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun readData() {
        try {
            val input: FileInputStream = openFileInput(fileName)
            val br = BufferedReader(InputStreamReader(input))
            val buffer = StringBuffer()
            var line: String?
            while (br.readLine().also { line = it } != null) {
                buffer.append(line).append("\n")
            }
            Log.d("MainActivity", buffer.toString())
            binding.tvData.text = buffer.toString()
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun saveDataByCache() {
        val fileOutputStream: FileOutputStream
        val file: File
        try {
            file = File(cacheDir, fileName)
            Log.d("MainActivity", cacheDir.absolutePath)
            fileOutputStream = FileOutputStream(file)
            fileOutputStream.write(content.toByteArray())
            fileOutputStream.close()
            Toast.makeText(this, "Save data successfully", Toast.LENGTH_SHORT).show()
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    override fun readDataByCache() {
        try {
            val file = File(filesDir, fileName)
            val br = BufferedReader(InputStreamReader(FileInputStream(file)))
            val buffer = StringBuffer()
            var line: String?
            while (br.readLine().also { line = it } != null) {
                buffer.append(line).append("\n")
            }
            Log.d("MainActivity", buffer.toString())
            binding.tvData.text = buffer.toString()
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    companion object {
        private const val fileName = "vanphuc0503.com"
        private const val content = "Noi dung duoc chia se thong qua internal storage v2"
    }
}